src/Shared/CopyOnWriteDictionary.cs

```diff
@@ -174,112 +153,29 @@ public CopyOnWriteDictionary(IDictionary<K, V> dictionary)
-                }
-
-                return _isDummy;
-            }
-        }
-
-        private bool _isDummy;
-
         /// <summary>
         /// Comparer used for keys
         /// </summary>
-        internal IEqualityComparer<K> Comparer { get; private set; }
-
-        /// <summary>
-        /// Gets the backing dictionary for reading.
-        /// </summary>
-        private CopyOnWriteBackingDictionary<K, V> ReadOperation
+        internal IEqualityComparer<K> Comparer
         {
-            get
-            {
-                ErrorUtilities.VerifyThrow(!IsDummy || backing == null || backing.Count == 0, "count"); // check count without recursion
-#if DEBUG
-                if (s_forceWrite)
-                {
-                    if (!IsDummy)
-                    {
-                        return WriteOperation;
-                    }
-                }
-#endif
-                if (backing == null)
-                {
-                    return CopyOnWriteBackingDictionary<K, V>.ReadOnlyEmptyInstance;
-                }
-
-                return backing;
-            }
-        }
-
-        /// <summary>
-        /// Gets the backing dictionary for writing.
-        /// </summary>
-        private CopyOnWriteBackingDictionary<K, V> WriteOperation
-        {
-            get
-            {
-                ErrorUtilities.VerifyThrow(!IsDummy, "dummy");
-
-                if (backing == null)
-                {
-                    backing = new CopyOnWriteBackingDictionary<K, V>(capacity, Comparer);
-                }
-                else
-                {
-                    lock (((ICollection)backing).SyncRoot)
-                    {
-                        backing = backing.CloneForWriteIfNecessary();
-                    }
-                }
-
-                return backing;
-            }
+            get => _backing.KeyComparer;
+            private set => _backing = _backing.WithComparers(keyComparer: value);
         }
 
         /// <summary>
         /// Accesses the value for the specified key.
         /// </summary>
         public V this[K key]
         {
-            get => ReadOperation[key];
+            get => _backing[key];
 
             set
             {
-                if (!IsDummy)
-                {
-                    if (ReadOperation.HasNoClones)
-                    {
-                        WriteOperation[key] = value;
-                    }
-                    else
-                    {
-                        // Try to avoid a clone if it already is present with the same value
-                        if (!ReadOperation.TryGetValue(key, out V existingValue) || !EqualityComparer<V>.Default.Equals(existingValue, value))
-                        {
-                            WriteOperation[key] = value;
-                        }
-                    }
-                }
+                _backing = _backing.SetItem(key, value);
```

Comment:
Isn't this a data race on the contents of the `_backing` reference? The old implementation [used a lock](https://github.com/dotnet/msbuild/blob/master/src/Shared/CopyOnWriteDictionary.cs#L246) whenever it cloned the backing dict, but the new implementation does not lock anymore when it clones the backing dict. This means that multiple writers can race, right? The comments on the class also specify that multiple writers are made thread safe by locking the writes.

Reply:
_backing is immutable, so you should get a whole new dictionary every time you do this, preventing races.

Reply:
I was referring to the `_backing` reference, not the ImmutableDictionary object it refers to.

Here's the scenario:

- You have two reader threads: T1, and T2
- Both threads have a reference to the same CopyOnWriteDictionary instance: COWD1. COWD1._backing points to ImmutableDictionary ID1
- ID1 has the following items in it: `{p1: foo, p2: bar}`
- T1 and T2 both decide to become writer threads
  - T1 calls COWD1[p3] = p3Val
  - T2 calls COWD1[p4] = p4Val
- The following interleaving happens
  - T1 setter sets COWD1._backing to ID2: `{p1: foo, p2: bar, p3: p3Val}`
  - T2 setter sets COWD1._backing to ID3: `{p1: foo, p2: bar, p4: p4Val}`

Now we have a situation in which T1's update got lost. The reference to ID2 gets lost.

Reply:
Ah, misunderstood. That sounds legit.

Reply:
I don't believe that the old implementation provided this guarantee. From a comment on the old class:

```C#
    /// Thread safety: for all users, this class is as thread safe as the underlying Dictionary implementation, that is,
    /// safe for concurrent readers or one writer from EACH user. It achieves this by locking itself and cloning before
    /// any write, if it is being shared - i.e., stopping sharing before any writes occur.
```

Note the emphasis on EACH user. The way I read this, you are not safe to write to the data structure from multiple threads concurrently unless each thread operates on its own clone.

This seems to be supported by the implementation. If multiple threads attempt the operation as described above and the COW dictionary has not been cloned, they both attempt to write to the underlying `Dictionary<K, V>` which sure is not thread-safe. In fact, the first write access will even race on this assignment:

```C#
                if (backing == null)
                {
                    backing = new CopyOnWriteBackingDictionary<K, V>(capacity, Comparer);
                }
```

so the exact issue you outlined above where one of the writes may be lost can occur here as well.

Reply:
I've put together this test code and it indeed throws against the old implementation:

```C#
            Random random = new Random();
            CopyOnWriteDictionary<int, int> dd = new CopyOnWriteDictionary<int, int>();
            ThreadStart func = () =>
            {
                while (true)
                {
                    int key = random.Next();
                    dd[key] = random.Next();
                    if (!dd.ContainsKey(key))
                    {
                        throw new Exception("Race!");
                    }
                }
            };
            Thread th1 = new Thread(func);
            Thread th2 = new Thread(func);

            th1.Start(); th2.Start();
            th1.Join(); th2.Join();
```

Reply:
Aha, so an instance of COWD must be confined to one writer thread. Multiple writer threads must not use the same COWD instance. I guess I didn't make that connection from reading the class comments, I confused user with thread, and it wasn't clear whether "this dictionary" refers to the COWD type, or the wrapped dictionary :)

Under these constraints, yup, all looks good.